### PR TITLE
replace "rclcpp::ok()" by atomic-flag to avoid UAF bug

### PR DESCRIPTION
**Pull request by GoesM**
---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | #691   |
| Primary OS tested on | Ubuntu22.04 |
| Robotic platform tested on | gazebo simulation of Tally |

---

## Description of contribution in a few bullet points

- add an atomic-flag for threads

## Description of documentation updates required from your changes

- avoid UAF bug caused by `rclcpp::ok()` 's wrong work

---

### CODE REVIEW

File: src/slam_toolbox_common.cpp [hunk 1]
```diff
@@ -240,6 +241,8 @@ SlamToolbox::on_shutdown(const rclcpp_lifecycle::State & state)
 SlamToolbox::~SlamToolbox()
 /*****************************************************************************/
 {
+  stop_threads_.store(true);
```
Comment (SteveMacenski) [inline src/slam_toolbox_common.cpp:245]: Needs to be in the `on_deactivate` function as well

**Review by SteveMacenski — COMMENTED** [review]
Otherwise LGTM

File: src/slam_toolbox_common.cpp (hunk 1 above)
Comment (SteveMacenski) [reply src/slam_toolbox_common.cpp:245]: @GoesM just one thing to finish this up!

Comment (GoesM) [reply src/slam_toolbox_common.cpp:245]: I've added it .

File: src/slam_toolbox_common.cpp [hunk 2]
```diff
@@ -175,6 +176,7 @@ CallbackReturn SlamToolbox::on_deactivate(const rclcpp_lifecycle::State &)
 /*****************************************************************************/
 {
   RCLCPP_INFO(get_logger(), "Deactivating");
+  stop_threads_.store(true);
```
Comment (SteveMacenski) [inline src/slam_toolbox_common.cpp:179]: Sorry alsos `stop_threads_.store(false);` in activate so that it can be reactivated.

**Review by SteveMacenski — COMMENTED** [review]
Then can merge!

**Pull request closed without merging**